hypothesis-python/src/hypothesis/searchstrategy/strategies.py

```diff
@@ -624,6 +624,41 @@ def do_draw(self, data):
                 # As long as we consume data, we'll eventually pass or raise.
                 # But if we don't this could be an infinite loop.
                 assume(data.index > start_index)
+
+        # Special-case for SampledFrom strategy.  This is special even among
+        # strategies with a known and small(ish) list of possible elements,
+        # because it implements drawing directly as choice of an index.
+        # That means we can use the "make your own luck" trick explained in
+        # the shrinking guide to find even very rare allowed values.
+        from hypothesis.searchstrategy.lazy import LazyStrategy
+        from hypothesis.searchstrategy.misc import SampledFromStrategy
+
+        while isinstance(self.filtered_strategy, LazyStrategy):
+            self.filtered_strategy = self.filtered_strategy.wrapped_strategy
+
+        if isinstance(self.filtered_strategy, SampledFromStrategy):
```

Comment:
At first I was horrified to see this special case inserted directly into `FilteredStrategy`.

I'm actually warming up to it a bit, but a big drawback is still that there's no direct indication in `SampledFromStrategy` that it has this special handling.

An alternative might be to move most of this code into a special hook method in `SampledFromStrategy`, and then have `FilteredStrategy` call that method from here after a `hasattr` check.

That would retain most of the simplicity of this approach, while being less baffling to future readers of `SampledFromStrategy`.

Reply:
Extracting this to a method would worry me, since calling it from anywhere else could very easily break some important invariants (e.g. the previous block must be from `cu.choice(data, self.elements)`).  It's a good point that `SampledFromStrategy` should show that the special handling exists though, so I've added a big comment block.

Reply:
That's fair.

Part of me wants to make this a more general mechanism, but realistically that isn't worthwhile until we actually have a second or third use-case, which might never happen.

So for the foreseeable future I'm OK with a hard-coded exception and a suitable comment in `SampledFromStrategy`.